### PR TITLE
PR review

### PR DESCRIPTION
**Pull request by zenoma**

### CODE REVIEW

File: .env
Comment (davidbarral) [inline .env:1]: Normally you would ignore the `.env` file in the repo and add a `.env.template` file with an example and some comments about the meaning of each variable. The `.env` file is supposed to be configured per user locally, so having it versioned will cause it to be marked as modified so you force the user to use a `git update-index --assume-unchanged` to avoid false positives. But, using that will make difficult to maintain this file when you need to add a new var (forcing the user to do a `git add --patch` and be very aware of what must change and what must not.  

To summarize.

- Remove .env from the repo
- Add an .env.template instead
- Add in the README an step to setup the template. Even better add a `./bin/setup` script that can be run on `postinstall` in the `package.json` an it is idempotent.


**THESE FIXES ARE OPTIONAL**

File: authServer.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+const express = require('express');
+const dotenv = require('dotenv');
```
Comment (davidbarral) [inline authServer.js]: Why use `dotenv`. Latest node versions have an `--env` support which is more than enough for this. Using dotenv seems a bit overkill and it's always better to use the vanilla version of things if possible. 

To improve the DX maybe adding some scripts to run the project would be helpful. 

```
npm run start:auth-server
npm run start:service
npm run client
```

Also to be even more friendlier, a `npm start`  script would be nice. This script should run the service and the auth-server concurrently (there are several tools that do this).  Also running the servers with the node `--watch` mechanism would be great to allow server autoreloading when the code changes. You could use some tool like `nodemon` but, it's the same as using dotenv, a bit overkill and I'd rather go the vanilla way. 

Also regarding node, there is no indication in the README about which node version I should use. It would be desirable to have an `engines` entry. 
Also it would be great to have a `.node-version` file with the version you are using. It's widely supported.

- Add engines to the package.json
- Add a .node-version file
- Add info in the README about the node version
- Remove dotenv, use `--env`.
- Add some `package.json` scripts.
- Use some watch mode.

**THESE FIXES ARE OPTIONAL**.

Another approach to solve all this stuff would be using Docker and a docker compose file to run all the services.

File: authServer.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const express = require('express');
```
Comment (davidbarral) [inline authServer.js]: As of today I don't see value in starting a new project using commonJS. New code should use the ESM standard. 

- Add `"type": "module"` to the package.json
- Switch from commonJS `require` to ESM `import`.

**THESE FIXES ARE OPTIONAL**

File: authServer.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+const dotenv = require('dotenv');
+const authRoutes = require('./routes/authRoutes');
+
+dotenv.config();
+
+const app = express();
```
Comment (davidbarral) [inline authServer.js]: Why express? What other alternatives are you aware of?

File: client.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+const dotenv = require('dotenv');
```
Comment (davidbarral) [inline client.js]: Same as above 😬

File: client.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+dotenv.config();
+
+const authServerURL = process.env.BASE_URL + ':' + process.env.AUTH_SERVER_PORT;
+const serviceURL = process.env.BASE_URL + ':' + process.env.SERVICE_PORT;
```
Comment (davidbarral) [inline client.js]: You are not setting up a linter/formatter so you want be advised to use string interpolation here. But you should prefer string interpolation to plain concatenation.

Also client wise, the env vars should be mor like  `AUTH_SERVER_URL` and `SERVICE_URL`.  Not going to be picky here with this xD

File: client.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+const loginCredentials = {
+  login: 'user@email.com',
+  password: '12345678',
+};
```
Comment (davidbarral) [inline client.js]: This should be env vars too. I know it's a dummy project but...

File: client.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+const getToken = async () => {
+  try {
+    console.log(new Date().toUTCString() + ' [Fetch token from Auth Server using credentials]');
```
Comment (davidbarral) [inline client.js]: I would have extracted this to a custom/simple logger object. The `new Date().toUTCString()` repeated all over is not very elegant 😬

File: client.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    const response = await fetch(`${authServerURL}/login`, {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(loginCredentials),
+    });
+
+
+    if (!response.ok) {
+      throw new Error( new Date().toUTCString() + ` [Failed to fetch token - ${response.statusText}]`);
```
Comment (davidbarral) [inline client.js]:
```suggestion
      throw new Error(`Failed to fetch token - ${response.statusText`);
```
If you want to enrich the Error, use a custom Error subclass and add that as metadata, otherwise the error handling in the catch below will be very noisy.

File: client.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    throw error;
+  }
+};
+
+const getDataFromService = async (token) => {
+  try {
+    console.log(new Date().toUTCString() + ' [Fetch data from Service with the obtained token]');
+    const response = await fetch(serviceURL, {
+      method: 'GET',
+      headers: { Authorization: token },
```
Comment (davidbarral) [inline client.js]: Nitpicking here, but OAuth like solutions tend to use the standarr `Bearer <token>` format for the Authorization headers. It's not mandatory of course as long as you control all the pieces, but, it's nice to do so.

File: client.js [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    const data = await response.json();
+    console.log(data);
+  } catch (error) {
+    console.error(error.message);
+  }
+};
+
+const main = async () => {
```
Comment (davidbarral) [inline client.js]: **OPTIONAL EXERCISE**

Cache the token and use the `exp` field to renew the token on demand. Make sure that you handle race conditions 🤓.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,14 @@
+{
+  "dependencies": {
+    "dotenv": "^16.3.1",
+    "express": "^4.18.2",
+    "jsonwebtoken": "^9.0.2",
+    "node-fetch": "^3.3.2"
```
Comment (davidbarral) [inline package.json]: You are not using this. Remove it. Your code is using the `global.fetch` function. Present in the latest versions of node (built as a wrapper of the `undici` client).

If you were using this your code would fall apart, because you'll be doing a `require("node-fetch")` and since version 3.x this package is ESM only and can only be imported in an ESM module or with a dynamic `import` in a commonJS project.

File: package.json [hunk 12]
```diff
@@ -0,0 +1,14 @@
...
+  },
+  "devDependencies": {
+    "jest": "^29.7.0"
+  },
+  "scripts": {
+    "test": "jest"
```
Comment (davidbarral) [inline package.json]: Nowadays I would recommend using vitest. Wich is for the most par a drop in replacement for Jest but it's based on esbuild instead of babel and is waaaay faster an less problematic.

File: middleware/jwtMiddleware.js
Comment (davidbarral) [inline middleware/jwtMiddleware.js:1]: This is not a middleware. It's two utility functions. Bad naming is misleading (it mislead me, in fact xD).

File: middleware/jwtMiddleware.js [hunk 13]
```diff
@@ -0,0 +1,18 @@
+const jwt = require('jsonwebtoken');
+const dotenv = require('dotenv');
+dotenv.config();
+
+const secret_phrase = process.env.SECRET_PHRASE;
```
Comment (davidbarral) [inline middleware/jwtMiddleware.js]:
```suggestion
const secretPhrase = process.env.SECRET_PHRASE;
```

Be consistent with the case.

File: middleware/jwtMiddleware.js [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+
+const signJWT = (data) => {
+  return jwt.sign(data, secret_phrase);
+};
+
+const verifyJWT = (token, callback) => {
+  jwt.verify(token, secret_phrase, callback);
```
Comment (davidbarral) [inline middleware/jwtMiddleware.js]: My two cents here. Do not use callbacks, use a promise. Don't be afraid of making this function async. Make the promise return true/false.

File: middleware/jwtMiddleware.js [hunk 15]
```diff
@@ -0,0 +1,18 @@
+const jwt = require('jsonwebtoken');
+const dotenv = require('dotenv');
+dotenv.config();
+
+const secret_phrase = process.env.SECRET_PHRASE;
+
+const signJWT = (data) => {
+  return jwt.sign(data, secret_phrase);
```
Comment (davidbarral) [inline middleware/jwtMiddleware.js]: I'm missing here passing the `expiresIn` option to set an expiration time for the token (and an env var to set said expiration time).

File: middleware/userValidationMiddleware.js
Comment (davidbarral) [inline middleware/userValidationMiddleware.js:1]: Again, this is not a middleware, it's an utility function.

File: routes/serviceRoutes.js [hunk 16]
```diff
@@ -0,0 +1,25 @@
+const express = require('express');
+const jwtMiddleware = require('../middleware/jwtMiddleware');
+
+const router = express.Router();
+
+router.get('/', (req, res) => {
+
+    const token = req.headers.authorization;
```
Comment (davidbarral) [inline routes/serviceRoutes.js]: The thing about `Bearer <token>` here too.

File: routes/serviceRoutes.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+          return res.status(401).json({ message: 'Invalid Token' });
+        }
+  
+        console.log(new Date().toUTCString() + ' [Valid token usage for ' + decoded.login + ']')
+        res.json({ message: "Data fetched", user: decoded.login });
+      });
+    } else {
+      res.status(401).json({ message: 'Token not provided' });
+    }
+  });
```
Comment (davidbarral) [inline routes/serviceRoutes.js]: - Using a promise in `verifyJWT` will make this code much readable.
- I would change the `verifyJWT` function to resolve early with `false` if no token is provided, here you would only deal with two cases: ok token and bad token. 
- Truth be told this router here should assume the token is correct if its's run and you should use a real middleware to control the token.

File: tests/authServer.test.js
Comment (davidbarral) [inline tests/authServer.test.js:1]: I think it's great that you are doing integration tests by using fetch request against the real services, but it would be nice to make the test launch the services on their own (if they are not running already), and stop them after finishing. Think about running the tests on CI. How would you do that? What problems could you expect?

File: tests/authServer.test.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+    headers: {
+      'Content-Type': 'application/json',
+    },
+    body: JSON.stringify(credentials),
+  });
+
+  const responseBody = await response.json();
+
+  expect(response.status).toBe(200);
+  expect(responseBody.token).toBeTruthy();
```
Comment (davidbarral) [inline tests/authServer.test.js]: I would expect it to be a real jwt token. Check for some trait. This test pases if your service returns anything that's not `false`, `0` or `""`.

File: tests/authServer.test.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
+const dotenv = require('dotenv');
+dotenv.config();
+
+const authServerURL = process.env.BASE_URL + ':' + process.env.AUTH_SERVER_PORT;
+
+test('givenValidData_whenLogin_then200AndValidToken', async () => {
```
Comment (davidbarral) [inline tests/authServer.test.js]: Did you use some code generator for this?

File: tests/authServer.test.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
+const dotenv = require('dotenv');
```
Comment (davidbarral) [inline tests/authServer.test.js]: You do this on each test... maybe you should do it instead in a jest setup file.

**Review by davidbarral — CHANGES_REQUESTED** [review]
Generally speaking it's ok. It does the job.

Notes:

- Layout wise I would have put each piece code in its own folder: `service`, `auth-server`, `client`.
- I would setup a minimal formatter at least. Using prettier is almost trivial. Setting a linter is a bit more work, but if you use biome.js you can have both with a very nice set of defaults.

File: authServer.js (hunk 3 above)
Comment (zenoma) [reply authServer.js]: I choosed Express because it's a simple and straightforward project and, I've already worked with it in another example project. I've heard of a couple of other server frameworks, but I haven't really work with them yet.

File: tests/authServer.test.js (hunk 19 above)
Comment (zenoma) [reply tests/authServer.test.js]: No, I've used this naning convention on my TFG and in some other assignments. I like the explicity of it, because you don't have to search in the code to know what is being tested.

File: tests/authServer.test.js
Comment (zenoma) [reply tests/authServer.test.js:1]: I did that tests just to make an example of suit case with unit tests on this services. A good approach to test this services effectively would have a combination of unit tests, integration tests, and end-to-end tests. The CI pipeline for the end-to-end tests could utilize a coverage reporting tool, integrated with GitHub Actions or Jenkins, to generate coverage reports. If we dont want to depend on Github CI/CD, we could build a full CI/CD with Jenkins, SonarQube, Docker and Kubernetes, for example.

Some of the problems that could be expected include dependency management and dealing with different environments.

File: authServer.js (hunk 1 above)
Comment (zenoma) [reply authServer.js]: I used dotenv because didn't know that was a build-in solution. However, I can't make work the --env solution. I think is a problem node version, I'm currently using 18.13.0.

File: client.js (hunk 7 above)
Comment (zenoma) [reply client.js]: I didn't build a custom logger because I tought that I only needed to print a few messages.

File: package.json (hunk 11 above)
Comment (zenoma) [reply package.json]: This case was the opposite of the 'dotenv', I started with node-fetch and discoreverd that node has now a proper fetch. Forgot to remove the dependency.

File: package.json (hunk 12 above)
Comment (zenoma) [reply package.json]: I understand now why Jest is more problematic, as soon as I removed the commonJS imports Jest problems appeared, like not being compatible wiht ESM.


I will try to migrate the tests from Jest to Vitests.

File: package.json [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+  },
+  "scripts": {
+    "start:auth-server": "node ./auth-server/authServer.js &",
+    "start:service": "node ./service/service.js &",
+    "start:auth-server:watch": "node --watch ./auth-server/authServer.js",
+    "start:service:watch": "node --watch ./service/service.js",
+    "client": "node ./client/client.js",
+    "test": "npm run start:service && npm run start:auth-server && vitest run && npm run kill:service && npm run kill:auth-server",
+    "kill:service": "pkill -f ./service/service.js",
+    "kill:auth-server": "pkill -f ./auth-server/authServer.js'"
```
Comment (davidbarral) [inline package.json:22]: OK. But two things:

1. Notice that if you use pkill your are killing the process by name. If you have two sessions running the same tests or the servers up for developmet you'll have undesired results
2.   It's better if the test suite that spins up the services and stops them as needed. The test suite can store the pid of the forked server and kill it by pid.

File: .eslintrc [hunk 22]
```diff
@@ -0,0 +1,3 @@
+{
+    "extends": ["biome:recommended"]
+}
```
Comment (davidbarral) [inline .eslintrc:3]: Maybe I'm missing something but it seems that you are not defining the necessary dependencies for this config.

And if you are going to use biome rules... why don't use biome and drop eslint and prettier. :D

File: Biomefile
Comment (davidbarral) [inline Biomefile:1]: I don't have a clue about what this file does xD

File: utils/simpleLogger.js [hunk 23]
```diff
@@ -0,0 +1,9 @@
+class SimpleLogger {
```
Comment (davidbarral) [inline utils/simpleLogger.js:1]: Not a fan of using classes, but fair enough :D

File: service/service.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
+import express from 'express';
+import dotenv from 'dotenv'
+
+import serviceRoutes from './routes/serviceRoutes.js';
+import SimpleLogger from '../utils/simpleLogger.js';
+
+const logger = new SimpleLogger();
+dotenv.config();
```
Comment (davidbarral) [inline service/service.js:8]: This should be the first thing after the imports. Moreover, given that the ESM imports are resolved statically, if you want to make sure that any module can access the env variables, you should move the dotenv import and this call to a separate module/file and make it the first import.

File: service/tests/service.test.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
+import { expect, test } from 'vitest'
+
+import dotenv from 'dotenv'
+
+dotenv.config();
```
Comment (davidbarral) [inline service/tests/service.test.js:5]: This would be better located in a vite setup file and not in the test, unless you want each test to be able to load a different .env file (but I don't think that's the case here).

File: service/tests/service.test.js [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+      login: 'user@email.com',
+      password: '12345678',
+    };
+  
+    const authResponse = await fetch(`${authServerURL}/login`, {
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify(credentials),
```
Comment (davidbarral) [inline service/tests/service.test.js:22]: A a side note, no fetch should be issued without a timeout (every fetch in the code. I mean, not just this one).

```
fetch(url, {
   // options...
   signal: AbortSignal.timeout(2000)
})
```

It's not too problematic in test suites because vite timeouts the test execution (after 4 seconds maybe... 🤔)

**Review by davidbarral — APPROVED** [review]

File: package.json (hunk 21 above)
Comment (zenoma) [reply package.json:22]: I didn't realize that could happen. Next time this could be a good practice to do, thanks.

File: Biomefile
Comment (zenoma) [reply Biomefile:1]: I gave implementing Biome a shot for the first time. I thought that it was fine, but it seems that is not the way to do it.

File: service/tests/service.test.js (hunk 25 above)
Comment (zenoma) [reply service/tests/service.test.js:5]: Like you said to me I tried to use the --env variable and get rid of the _dotenv_ dependency but didn't make it to work, I think it was for my nodeJs version. So I left all of that and didn't change it.
